torch/csrc/jit/tensorexpr/loopnest.cpp

```diff
@@ -759,6 +759,97 @@ void LoopNest::computeInlineWithRandom(Stmt* s) {
   inlined_random_functions_.insert(stmt_to_tensor_.at(s)->function());
 }
 
+// TODO: Unify with DepTracker
+class UseFinder : public IRVisitor {
+ public:
+  std::unordered_map<const Buf*, std::vector<BufUse>> findUses(Stmt* s) {
+    uses_.clear();
+    s->accept(this);
+    return uses_;
+  }
+
+ private:
+  void visit(const Store* v) override {
+    if (stores_[v->buf()].insert(last_stmt_).second) {
+      uses_[v->buf()].push_back({(Stmt*)v, true});
+    }
+    last_stmt_ = (Stmt*)v;
+    IRVisitor::visit(v);
+  }
+  void visit(const Load* v) override {
+    if (loads_[v->buf()].insert(last_stmt_).second) {
+      uses_[v->buf()].push_back({last_stmt_, false});
+    }
+    IRVisitor::visit(v);
+  }
+
+  Stmt* last_stmt_ = nullptr;
+  std::unordered_map<const Buf*, std::vector<BufUse>> uses_;
+
+  // Sets of loads and stores in order to keep the results unique
+  std::unordered_map<const Buf*, std::unordered_set<Stmt*>> loads_;
+  std::unordered_map<const Buf*, std::unordered_set<Stmt*>> stores_;
+};
+
+std::unordered_map<const Buf*, std::vector<BufUse>> findUses(Stmt* s) {
+  UseFinder uf;
+  return uf.findUses(s);
+}
+
+class ContainedStmtsFinder : public IRVisitor {
+ public:
+  // Simply list all Stores and LetStmts that are children of the given stmt
+  std::unordered_set<Stmt*> findContainedStmts(Stmt* s) {
+    contained_.clear();
+    s->accept(this);
+    return contained_;
+  }
+
+ private:
+  void visit(const Store* v) override {
+    contained_.insert((Stmt*)v);
+    IRVisitor::visit(v);
+  }
+  void visit(const LetStmt* v) override {
+    contained_.insert((Stmt*)v);
+    IRVisitor::visit(v);
+  }
+
+  std::unordered_set<Stmt*> contained_;
+};
+
+bool containsAll(const std::vector<BufUse>& uses, Block* b) {
+  std::unordered_set<Stmt*> not_found;
+  for (auto use : uses) {
+    not_found.insert(use.s);
+  }
+
+  ContainedStmtsFinder csf;
+  std::unordered_set<Stmt*> contained = csf.findContainedStmts(b);
+  for (auto s : contained) {
+    not_found.erase(s);
+  }
+  return not_found.empty();
+}
+
+Block* findParentBlock(Stmt* s) {
+  while (s) {
+    if (auto b = dynamic_cast<Block*>(s)) {
+      return b;
+    }
+    s = s->get_parent();
+  }
+  return nullptr;
+}
+
+Block* findLowestContainingBlock(const std::vector<BufUse>& uses) {
```

Comment:
Minor: add a TODO to see if we should improve the performance of this block for larger programs. Many systems I knew look at all the parents in reverse order, and find the lowest one.